vm-virtio/src/mem.rs

```diff
@@ -0,0 +1,951 @@
+ * - VIRTIO_MEM_REQ_STATE
+ */
+const VIRTIO_MEM_RESP_ERROR: u16 = 3;
+
+/* State of memory blocks is "plugged" */
+const VIRTIO_MEM_STATE_PLUGGED: u16 = 0;
+/* State of memory blocks is "unplugged" */
+const VIRTIO_MEM_STATE_UNPLUGGED: u16 = 1;
+/* State of memory blocks is "mixed" */
+const VIRTIO_MEM_STATE_MIXED: u16 = 2;
+
+/* request to plug memory blocks */
+const VIRTIO_MEM_REQ_PLUG: u16 = 0;
+/* request to unplug memory blocks */
+const VIRTIO_MEM_REQ_UNPLUG: u16 = 1;
+/* request to unplug all blocks and shrink the usable size */
+const VIRTIO_MEM_REQ_UNPLUG_ALL: u16 = 2;
+/* request information about the plugged state of memory blocks */
+const VIRTIO_MEM_REQ_STATE: u16 = 3;
+
+// Get resize event.
+const RESIZE_EVENT: DeviceEventT = 0;
+// New descriptors are pending on the virtio queue.
+const QUEUE_AVAIL_EVENT: DeviceEventT = 1;
+// The device has been dropped.
+const KILL_EVENT: DeviceEventT = 2;
+// The device should be paused.
+const PAUSE_EVENT: DeviceEventT = 3;
+
+#[derive(Debug)]
+pub enum Error {
+    /// Guest gave us bad memory addresses.
+    GuestMemory(GuestMemoryError),
+    /// Guest gave us a write only descriptor that protocol says to read from.
+    UnexpectedWriteOnlyDescriptor,
+    /// Guest gave us a read only descriptor that protocol says to write to.
+    UnexpectedReadOnlyDescriptor,
+    /// Guest gave us too few descriptors in a descriptor chain.
+    DescriptorChainTooShort,
+    /// Guest gave us a buffer that was too short to use.
+    BufferLengthTooSmall,
+    /// Guest sent us invalid request.
+    InvalidRequest,
+    /// Failed to EventFd write.
+    EventFdWriteFail(std::io::Error),
+    /// Failed to EventFd try_clone.
+    EventFdTryCloneFail(std::io::Error),
+    /// Failed to MpscRecv.
+    MpscRecvFail(mpsc::RecvError),
+    /// Resize invalid argument
+    ResizeInval(String),
+    /// Fail to resize trigger
+    ResizeTriggerFail(DeviceError),
+}
+
+/// Got from qemu/include/standard-headers/linux/virtio_mem.h
+/// rust union doesn't support std::default::Default that
+/// need by mem.read_obj.
+/// Then move virtio_mem_req_plug, virtio_mem_req_unplug and
+/// virtio_mem_req_state to virtio_mem_req.
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Default)]
+struct VirtioMemReq {
+    req_type: u16,
+    padding: [u16; 3],
+    addr: u64,
+    nb_blocks: u16,
+}
+
+// Safe because it only has data and has no implicit padding.
+unsafe impl ByteValued for VirtioMemReq {}
+
+/// Got from qemu/include/standard-headers/linux/virtio_mem.h
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Default)]
+struct VirtioMemRespState {
+    state: u16,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Default)]
+struct VirtioMemResp {
+    resp_type: u16,
+    padding: [u16; 3],
+
+    state: VirtioMemRespState,
+}
+
+// Safe because it only has data and has no implicit padding.
+unsafe impl ByteValued for VirtioMemResp {}
+
+/// Got from qemu/include/standard-headers/linux/virtio_mem.h
+#[repr(C, packed)]
```

Comment:
This part use #[repr(C, packed)] because all of virtio device config use #[repr(C, packed)].
Other structs just work OK with #[repr(C)].
I think the reason is that they have different conversion methods.